### ISSUE TITLE
[IDEA] Prevent double rendering when $:/core/ui/PageTemplate is open in the story

### ISSUE THREAD

**Issue by saqimtiaz** [Jermolene/TiddlyWiki5#5136]
It can be quite disconcerting when a user has `$:/core/ui/PageTemplate` open in the story, that the page appears to have a double rendering and in some cases the sidebar becomes unusable. In particular this is a barrier to users that want to edit and customize the page template in the browser via the TiddlyWiki UI:

See for example this report:
https://github.com/Jermolene/TiddlyWiki5/commit/6c98bb706a15b353a92895529a69bece11f613ff#commitcomment-44594409

We should consider hiding/preventing the rendering of the `$:/core/ui/PageTemplate` tiddler when open in the story. One option would be to add the field `hide-body` with a value of `yes`.

**Comment by Jermolene** [comment]
Hi @saqimtiaz I've been thinking about that too. While the double sidebar issue is unique to the page template, viewing many core templates is pretty unhelpful at present.  One potential solution is a generic mechanism for marking tiddlers so that the view template displays them with a different template.

The "obvious" solution would be an optional "view-body-template" field that is used instead of the default body template. I don't like the way that it ties tiddlers directly to their template. An alternative might be a new "view-role" field that contains a token that is looked up via a prefix to obtain the title of the template. Then we might define roles for, say, templates and filters with special affordances for working with those data types.

**Comment by saqimtiaz** [comment]
> An alternative might be a new "view-role" field that contains a token that is looked up via a prefix to obtain the title of the template. Then we might define roles for, say, templates and filters with special affordances for working with those data types.

I like this idea. It would allow for instance, filter tiddlers to be displayed with the filter visualization tool we have been discussing.

I'll pick this thread up again after 5.1.23.

**Comment by joshuafontany** [comment]
This is kind of what I was aiming for with PR https://github.com/Jermolene/TiddlyWiki5/pull/4681

I don't think this is the way to go (but it was a fun experiment). Still, having a way to customize the view and edit templates without relyting on a field on each tiddler would be excellent to have available in the core.

**Comment by Jermolene** [comment]
> This is kind of what I was aiming for with PR https://github.com/Jermolene/TiddlyWiki5/pull/4681
> 
> 
> 
> I don't think this is the way to go (but it was a fun experiment). Still, having a way to customize the view and edit templates without relying on the limited 'hide-text' field on each tiddler functionality would be excellent to have available in the core.

Yes indeed @joshuafontany I should have referenced that PR. The other thing I hope to get done for v5.1.24 is a new flexgrid-based layout that supports multiple story views. One could imagine wanting to use different view templates for different story rivers (eg my own notes are in one column as skinny as possible, stuff I've imported is in another column with lots of visible metadata about where it came from).

We've already discussed breaking up the navigator widget into customisable action strings, perhaps #4681 prompts us to try to do something similar with the view/edit logic that is currently special-cased into the list widget.

**Comment by saqimtiaz** [comment]
> The other thing I hope to get done for v5.1.24 is a new flexgrid-based layout that supports multiple story views. One could imagine wanting to use different view templates for different story rivers (eg my own notes are in one column as skinny as possible, stuff I've imported is in another column with lots of visible metadata about where it came from).

@Jermolene I would be happy to help with testing/refining the flexgrid-based layout, as well as making sure that my Stories plugin is rewritten for it. That would serve as a good example of leveraging the flexibility of the layout. 

My Stories plugin does support different view templates for each story river btw, but honestly everything about the CSS/layout is currently kludgy by necessity and I very much look forward to a more flexible base layout.

**Comment by Jermolene** [comment]
> @Jermolene I would be happy to help with testing/refining the flexgrid-based layout, as well as making sure that my Stories plugin is rewritten for it. That would serve as a good example of leveraging the flexibility of the layout. 

I'll publish what I've got as soon as we get v5.1.23 out. One part I haven't addressed yet is the UI to let users add/remove columns, and also to set their width, and grow/shrink behaviour.

> My Stories plugin does support different view templates for each story river btw, but honestly everything about the CSS/layout is currently kludgy by necessity and I very much look forward to a more flexible base layout.

Yes indeed, I've spent far too long wrestling with the current layout. It'll be fun to work on trying to get things neat and flexible.

By the way, one big consequence of using flexgrid is that I haven't yet figured out how to make dropdowns escape the bounds of their columns. I fear we may need to introduce a new dropdown mechanism based on the modal mechanism.

**Comment by Jermolene** [comment]
I wondered about picking up a trick from the way that we [customise the naming of tiddler files under Node.js](https://tiddlywiki.com/prerelease/#Customising%20Tiddler%20File%20Naming), and having a cascade of filters that are applied in turn until one of them returns a non-empty result, which is then used as the template for the tiddler.

We would need a new operator `mapfirst` that takes a list of filters and returns the first non-empty result. For example:

```
\define filter1() [tag[mytag]then[$:/my-tag-template]]
\define filter2() [tag[anothertag]then[$:/another-tag-template]]
<$text text={{{ [<filter1>] [<filter2>] +[mapfirst[My Tiddler Title]] }}}/>
```

For the view template, the filter list would be driven by a tag such as `$:/tags/TemplateMapper`. The core would include some mappers for handling view vs edit mode:

```
[is[draft]then[$:/core/ui/EditTemplate]]
[[$:/core/ui/ViewTemplate]]
```

With careful use of "list-before" and "list-after" fields, plugin authors should be able to add their own templates for their own types of tiddlers.

I'm not sure whether this is the right approach:

* It only gives us swapping in and out of a singular view template, it doesn't help us splice custom content into the existing templates as does #5144 
* But we can apply this same pattern repeatedly, and have the main view template, body template and other templates all controlled separately
* `mapfirst` does quite a lot of work, which may have performance implications
* But I really like the way that this idea could fix the kludge of the list widget switching between view and edit templates

**Comment by joshuafontany** [comment]
I really appreciate how internally consistent yet flexible the approach to $:/config/FileSystemPath filters is. I had zero problems copying it to use as $:/config/FileSystemExtensions in my filesystemAdaptor PR. I like the idea of reusing this pattern or writing a core $tw.utils. method to build into a filter operator.

**Comment by rmunn** [comment]
>     * It only gives us swapping in and out of a singular view template, it doesn't help us splice custom content into the existing templates as does #5144

I think this is the wrong PR number, as #5144 is only a documentation PR and makes no code changes. I've been unable to figure out which PR or issue you meant here.

**Comment by saqimtiaz** [comment]
I think #4681

**Comment by Jermolene** [comment]
Yes sorry, I was indeed thinking of #4681, I've updated my post above

**Comment by AnthonyMuscio** [comment]
I have not followed every branch of this discussion, however please consider this.

I have commonly used in code tiddlers something like the following;
```
<$list filter="[all[current]!match[specific-tiddlername]]" variable=nul>
   Full tiddler content
</$list>
```
Then I do not see the result rendered when opening the code tiddler, because it can be misleading and cause the problem mentioned in the OT.

However I also have a showcode utility that hides the body and shows the un-rendered content (only) if so flagged. 

Perhaps a show code /hide code option on all system tiddlers would be the answer to this. We can then set the default for tiddlers such as $:/core/ui/PageTemplate

Happy to share my showcode macros if interested.

Tony

**Comment by Jermolene** [comment]
Hi @AnthonyMuscio indeed, techniques like that are the best we have at the moment. Here we're exploring alternatives that avoid the need to modify a tiddler in order to change how it is displayed.

**Comment by AnthonyMuscio** [comment]
Currently the **hide-body=yes** field turns off the body viewtemplate, and an Alternate in the $:/tags/ViewTemplate can be used. 

If the body responded to a `<<hide-body>>` variable then the viewTemplate could also hide the default display of the tiddler content. This variable can be set like a visibility indicator, in an external state tiddler.

Tony

**Comment by joshuafontany** [comment]
> I really appreciate how internally consistent yet flexible the approach to $:/config/FileSystemPath filter handling is. I had zero problems copying it to use as $:/config/FileSystemExtensions in my filesystemAdaptor PR. I like the idea of reusing this pattern or writing a core $tw.utils. method to build into a filter operator.

@saqimtiaz @Jermolene @pmario This comment reminded me of something @saqimtiaz and I discussed during out brief Yjs realtime-multiuser-sync demo. We each have used this pattern for multiple things, and we talked about getting a core $tw.utils function (or similar) for this type of filtering.

I can definitely see using it for Customizable View Templates, for example a set of filters that return a list of ViewTemplate segments, and the first one to output a non-empty result wins. Then that is used as the $list to build the viewTemplate.

I also use it to implement some basic CORS handling, by using a set of filters that take the current http(S) request's Origin header, and running it past a series of filters. If a match is returned, the CORS Origin is whitelisted. I have that in one of my server PRs.

**Comment by saqimtiaz** [comment]
Yes, I use the FileSystemPath filters approach in my PUT uploader for Node.js and WebDAV, to determine the upload path and filename. However, adding such complexity to the uploaders goes against the model of uploaders being relatively easy to write and maintain. The pattern is useful enough that I feel it warrants  a core method that can be re-used.

A utils or wiki method along the lines of the following signature would be very helpful:

`resolveFilters(title,configTiddler,default)` or 

`resolveFilters(title,options)` where options is an object like `{configTiddler: "configTiddler", default:""}`

**Comment by Jermolene** [comment]
> A utils or wiki method along the lines of the following signature would be very helpful:

Yes indeed. The existing `generateTiddlerFilepath()` method could do with refactoring as it is pretty inefficient – it continues to evaluate the remaining filters after finding a result.

**Comment by Jermolene** [comment]
I've made a PR with the new "cascade" filter run prefix discussed above, and modifications to the view template to use the new prefix to choose the view template body template. See #6168.

**Comment by AnthonyMuscio** [comment]
> 
> 
> I've made a PR with the new "cascade" filter run prefix [discussed above](https://github.com/Jermolene/TiddlyWiki5/issues/5136#issuecomment-735435268), and modifications to the view template to use the new prefix to choose the view template body template. See #6168.

Jeremy, If I can be obvious I love the idea of a cascade but looking at your suggestion It is really hard to understand or follow (for me), I think you are bringing pre-knoweldge to the problem and mapfirst seems confusing (to me but I expect it will also to a large part of the community)

I have spent hours deep in contemplating the view/edit template mechanism of late, and it currently confuses the possibility of customising it. In fact it makes it hard to explain. If I may try though;

Whist the view template is driven by tags, it is about transcluding tiddlers. Unfortunately this complicates customisations. 
- The first issue is hiding $:/core/ui/ViewTemplate/body to use another 
- The second is the process by which we select another body or default to the existing.
- The broader problem is to allow a programaticaly driven selection of any template element such as replace the toolbar for some tiddler types
- All this is complicated because we usually need to fall back to an existing **view/edit template element** such as the title/subtitle and many existing solutions insert themselves into the template via a tag and a list-before or after.

My thoughts although hard to put in writing is developing a method to combine a tagging method (list not tagging of core tiddlers) with a the lookup operator and a list, to allow a true cascade in the UI template mechanism.
- Why the lookup? Because its result is a transclusion
- Why the tag because it selectively indicates which elements to use and the order
- Why a list because rather than a tag it can be stored elsewhere and not modify core tiddlers. This may also provide a method to maintain backward compatibility because tagging effectively updates a list.

The issues is also retaining backwards compatibility.

You will see this effort in my questions about using custom user/type in the type field, which for a while seemed a good approach.

I wonder if what we need is the ability to use a set of lists in a cascade, to extract the template items to actually display.

I do not yet have the answers but they exist in an unrealised part of my intuition so I apologise for not being able to voice it in black and white.

I would like to see a beautiful generic cascading tiddler title method (transclusion option) that would solve this.

**Comment by AnthonyMuscio** [comment]
Ok, Please try this POC on tiddlywiki.com (it only lists tiddler titles, in a working example it would be in the viewTemplate mechanism use the lookup operator to transclude the view template tiddler.

- First add an additional viewTemplate item using the tag $:/tags/ViewTemplate
- Then use the tag pill to change the order (this updates the list field)
- Then look at the below tiddler. 

This example uses the type field to determine the prefix for view templates but defaults to $:/core/ui/ViewTemplate/ but it also honours other tiddlers tagged $:/tags/ViewTemplate
- Use the edit button as an example to create an alternate template for type="user/type"
- You will see the alternate  viewTemplate is used instead, but will default to the normal one if not available

```

<<tag $:/tags/ViewTemplate>>

Current list =List {{$:/tags/ViewTemplate!!list}}
<hr>
<$set name=type value="user/type">
<$set name=type-prefix filter="[[$:/types/]] [<type>] [[/viewTemplate/]] +[join[]]">
type=<<type>> <<type-prefix>><br>
<$list filter="[{$:/tags/ViewTemplate!!list}enlist-input[]]" variable=template-item counter=item>
   <$list filter="[<template-item>prefix[$:/core/ui/ViewTemplate/]removeprefix[$:/core/ui/ViewTemplate/]]" variable=template-element>
        <<item>> <<template-element>> {{{ [<template-element>addprefix<type-prefix>has[title]] ~[<template-element>addprefix[$:/core/ui/ViewTemplate/]] }}} <$button>
<$action-createtiddler $basetitle={{{ [<template-element>addprefix<type-prefix>] }}} >
   <$action-navigate $to=<<createTiddler-title>>/>
</$action-createtiddler>
Edit
</$button><br>
   </$list>
  <$list filter="[<template-item>!prefix[$:/core/ui/ViewTemplate/]]" variable=template-element>
        <<item>>  <<template-element>> *<br>
   </$list>

</$list>
</$set>
;* = Found in <<tag $:/tags/ViewTemplate>> only.
```

Notes
- The above example is only a two level cascade ie; if no custom view template item then use default
- The use of the type field can be changed (but need it?)
- The provision of an empty alternative items is effectively removing the item from the template
- We can still add items via regular addition to the $:/tags/ViewTemplate and they could be written to honour the type field (or not).
- A true cascade would allow more than one additional viewTemplate element to be tested for with the first being used.
- The ability to trigger an update to the $:/tags/ViewTemplate list field, when tagging a tiddler with $:/tags/ViewTemplate would be good.

**Comment by AnthonyMuscio** [comment]
So my above example would be an alternative way to address the view and edit templates allowing alternatives for any view OR Edit template including hiding the body. In this case it uses the type field but the ability to instead use a different field a filter or global variable would be nice, perhaps that is where the true cascade occurs.

**Comment by joshuafontany** [comment]
@AnthonyMuscio The PR that @Jermolene actually accimplishes this really well without having to rely on a tiddker-field (although you can trigger that).

Right now only the body template is determined by a `:cascade` filter.... but I really like how it works and can definitely see using this fir multiple places in the UI.

I hope after some good examples, it will be easier for users to define thier own body-type-filter & add it to the cascade with list-after or list-before fields.

**Comment by saqimtiaz** [comment]
I think the proposed approach is great and inherently more flexible than **anything** else that has been discussed.

 In practice it was extremely easy to setup a custom body template and a series of fallback options in the cascade. The best part is that it doesn't tie you to any particular way of determining a template, you can write a filter for any condition.

**Comment by Jermolene** [comment]
I think this can be closed now that #6168 is merged
